From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Bacon <zachbacon@vba-m.com>
Date: Sun, 29 Oct 2023 15:04:14 -0400
Subject: [PATCH] xinitrc.in.in: Ensure all various XDG_* etc. env vars are
 passed to DBus and systemd

Services (such as tumblerd and xfce4-notifyd) are started either by DBus
or by systemd, and need to know things such as XDG_CONFIG_HOME and
XDG_CACHE_HOME. While also ensuring xinitrc scripts can be sourced if it's necessary

Signed-off-by: Zach Bacon <zachbacon@vba-m.com>
---
 scripts/xinitrc.in.in | 45 ++++++++++++++++++++++++++++++++++++++++---
 1 file changed, 42 insertions(+), 3 deletions(-)

diff --git a/scripts/xinitrc.in.in b/scripts/xinitrc.in.in
index ff044edf..c5f783cc 100755
--- a/scripts/xinitrc.in.in
+++ b/scripts/xinitrc.in.in
@@ -78,11 +78,50 @@ cat /dev/null $XRESOURCES | xrdb -merge -
 # load local modmap
 test -r $HOME/.Xmodmap && xmodmap $HOME/.Xmodmap
 
-# if XAUTHLOCALHOSTNAME is not set in systemd user session, starting of xfce4-notifyd, DISPLAY etc. will fail
-if command -v systemctl >/dev/null 2>&1 && systemctl --user list-jobs >/dev/null 2>&1; then # user session is running
-  dbus-update-activation-environment --systemd XAUTHLOCALHOSTNAME=$XAUTHLOCALHOSTNAME
+# ensure both the DBus session bus and systemd user session (if running) has
+# all the env vars it needs to properly populate the environment of child
+# processes
+if command -v dbus-update-activation-environment >/dev/null 2>&1; then
+  if command -v systemctl >/dev/null 2>&1 && systemctl --user list-jobs >/dev/null 2>&1; then # user session is running
+    systemd_arg='--systemd'
+  fi
+
+  dbus-update-activation-environment $systemd_arg \
+    DESKTOP_SESSION \
+    XAUTHLOCALHOSTNAME=$XAUTHLOCALHOSTNAME \
+    XDG_CACHE_HOME \
+    XDG_CONFIG_DIRS \
+    XDG_CONFIG_HOME \
+    XDG_CURRENT_DESKTOP \
+    XDG_DATA_DIRS \
+    XDG_DATA_HOME \
+    XDG_MENU_PREFIX \
+    XDG_RUNTIME_DIR \
+    XDG_SEAT \
+    XDG_SEAT_PATH \
+    XDG_SESSION_CLASS \
+    XDG_SESSION_DESKTOP \
+    XDG_SESSION_ID \
+    XDG_SESSION_PATH \
+    XDG_SESSION_TYPE \
+    XDG_STATE_HOME
 fi
 
+# Source system vendor xinitrc scripts
+if [ -d /usr/share/defaults/etc/X11/xinit/xinitrc.d ]; then
+  for f in /usr/share/defaults/etc/X11/xinit/xinitrc.d/*; do
+    [ -x "$f" ] && . "$f"
+  done
+  unset f
+fi
+
+# source system xinitrc scripts
+if [ -d /etc/X11/xinit/xinitrc.d ]; then
+  for f in /etc/X11/xinit/xinitrc.d/*; do
+    [ -x "$f" ] && . "$f"
+  done
+  unset f
+fi
 
 # check if we start xfce4-session with ck-launch-session. this is only
 # required for starting from a console, not a login manager
